### PR TITLE
[WIP] Add precision modes and floatX wrapping in test_distributions.

### PR DESCRIPTION
**Pull request by kyleabeauchamp**
Before patches, cuda + float32 gives `FAILED (errors=43, failures=8)`.

After patches, cuda + float32 gives `FAILED (errors=44, failures=1)`.  CPU + float64 still gives 100% success after patch.

I'm not going to figure out the theano inf / testval stuff in a finite amount of time, so I think the best we can do is fix the other test failures one batch at a time.

I tried to be conservative in the precision reductions on the tests.  I don't think anything was unreasonable: 64 to 32 bit could easily kill 7 digits of precision IIRC, while I nixed at most 5 decimals.

### CODE REVIEW

File: pymc3/tests/test_distributions.py [hunk 1]
```diff
@@ -24,6 +25,15 @@
 import scipy.stats.distributions as sp
 import scipy.stats
 
+# Keep track of floating point mode for determining equality decimal cutoffs.
+DTYPE = floatX(1.0).dtype.name
```
Comment (twiecki) [inline pymc3/tests/test_distributions.py]: I think `theano.config.floatX` is more direct.

File: pymc3/tests/test_distributions.py [hunk 2]
```diff
@@ -415,7 +425,7 @@ def test_wald(self, value, mu, lam, phi, alpha, logp):
         with Model() as model:
             Wald('wald', mu=mu, lam=lam, phi=phi, alpha=alpha, transform=None)
         pt = {'wald': value}
-        assert_almost_equal(model.fastlogp(pt), logp, decimal=6, err_msg=str(pt))
+        assert_almost_equal(model.fastlogp(pt), logp, decimal=select_decimal(float64=6, float32=1), err_msg=str(pt))
```
Comment (twiecki) [inline pymc3/tests/test_distributions.py]: This errors with `float32>1`?

File: pymc3/tests/test_distributions.py (hunk 1 above)
Comment (kyleabeauchamp) [reply pymc3/tests/test_distributions.py]: Fixed in latest commit

File: pymc3/tests/test_distributions.py (hunk 2 above)
Comment (kyleabeauchamp) [reply pymc3/tests/test_distributions.py]: Yes IIRC.  I think this was returning a logp whose value was ~-1E5, so 6 decimal places is actually 11 digits of precision---totally unattainable on float32.

Comment (aseyboldt) [reply pymc3/tests/test_distributions.py]: Is there a good reason for using `assert_almost_equal` instead of `assert_allclose`? Checking rtol instead of decimals seems far more reliable. It is also suggested in the docstring of `assert_almost_equal`.

Comment (kyleabeauchamp) [reply pymc3/tests/test_distributions.py]: IMHO yes, a relative tolerance would be more appropriate in several cases like this.

IMHO we might as well make a separate issue because there's likely to be 10 more PRs worth of precision fixes to get GPU tests passing.

Comment (twiecki) [reply pymc3/tests/test_distributions.py]: @kyleabeauchamp I agree. But since assert_allclose might be a more elegant fix to the problem you're fixing here, we should implement that immediately.

**Comment by kyleabeauchamp** [comment]
OK, I made the switch to `allclose()` but I need to re-run the tests and adjust the tolerances on my GPU.

**Comment by twiecki** [comment]
@kyleabeauchamp Shouldn't the tolerances for CPU float32 and GPU float32 be identical?

**Comment by kyleabeauchamp** [comment]
Ideally, yes.  But I still want to re-run the tests on my GPU before merging.

**Comment by kyleabeauchamp** [comment]
I tried changing to allclose with a relative tolerance and realized that it may be more complicated.  The issue is that the tests probe a range of values, which are sometimes large and sometimes zero.  Based on the equation below, we can't use rtol for cases where the target (`desired`) is zero.  

```
atol + rtol * abs(desired)
```

TL;DR: I reverted back to the old function.  It works; we can revisit relative vs. absolute tolerance in the future.

**Comment by twiecki** [comment]
I think it's fine to move ahead here. What do you think @aseyboldt?

**Comment by aseyboldt** [comment]
If `desired` is zero, I think the right thing to do is to use atol. But I don't want to keep you busy with this boring stuff forever...
